pantab/__init__.py

```diff
@@ -14,3 +14,58 @@
+#
+# Directly using HyperAPI's C functions always was and still is discouraged and
+# unsupported by Tableu. In particular, Tableau will not be able to provide
+# official support for this hack.
+#
+# Because this is highly brittle, we try to make the error message as
+# actionable as possible and guide users in the right direction.
+
+api_incompatibility_msg = """
+pantab is incompatible with the version of Tableau Hyper API installed on your
+system. Please upgrade both `tableauhyperapi` and `pantab` to the latest version.
+If doing so does not fix this issue, please file an issue at
+https://github.com/innobi/pantab/issues mentioning the exact pantab and HyperAPI
+versions which triggered this error.
+"""
+
+
+def _check_signature(f, sig):
+    func_type = ffi.typeof(f)
+    assert func_type.kind == "function", f"expected a function, got {func_type.kind}"
+    assert func_type.cname == sig, f"expected '{sig}', got '{func_type.cname}'"
+    return f
+
+
+try:
+    from tableauhyperapi.impl.dll import lib, ffi
+
+    # fmt: off
+    libpantab.init(
+        _check_signature(lib.hyper_decode_date, "hyper_date_components_t(*)(uint32_t)"),
+        _check_signature(lib.hyper_encode_date, "uint32_t(*)(hyper_date_components_t)"),
+        _check_signature(lib.hyper_decode_time, "hyper_time_components_t(*)(uint64_t)"),
+        _check_signature(lib.hyper_encode_time, "uint64_t(*)(hyper_time_components_t)"),
+        _check_signature(lib.hyper_inserter_buffer_add_null, "struct hyper_error_t *(*)(struct hyper_inserter_buffer_t *)"),
+        _check_signature(lib.hyper_inserter_buffer_add_bool, "struct hyper_error_t *(*)(struct hyper_inserter_buffer_t *, _Bool)"),
+        _check_signature(lib.hyper_inserter_buffer_add_int16, "struct hyper_error_t *(*)(struct hyper_inserter_buffer_t *, int16_t)"),
+        _check_signature(lib.hyper_inserter_buffer_add_int32, "struct hyper_error_t *(*)(struct hyper_inserter_buffer_t *, int32_t)"),
+        _check_signature(lib.hyper_inserter_buffer_add_int64, "struct hyper_error_t *(*)(struct hyper_inserter_buffer_t *, int64_t)"),
+        _check_signature(lib.hyper_inserter_buffer_add_double, "struct hyper_error_t *(*)(struct hyper_inserter_buffer_t *, double)"),
+        _check_signature(lib.hyper_inserter_buffer_add_binary, "struct hyper_error_t *(*)(struct hyper_inserter_buffer_t *, uint8_t *, size_t)"),
+        _check_signature(lib.hyper_inserter_buffer_add_raw, "struct hyper_error_t *(*)(struct hyper_inserter_buffer_t *, uint8_t *, size_t)"),
+        _check_signature(lib.hyper_rowset_get_next_chunk, "struct hyper_error_t *(*)(struct hyper_rowset_t *, struct hyper_rowset_chunk_t * *)"),
+        _check_signature(lib.hyper_destroy_rowset_chunk, "void(*)(struct hyper_rowset_chunk_t *)"),
+        _check_signature(lib.hyper_rowset_chunk_field_values, "struct hyper_error_t *(*)(struct hyper_rowset_chunk_t *, size_t *, size_t *, uint8_t * * *, size_t * *, int8_t * *)"),
+    )
+    # fmt: on
+
+except Exception as e:
```

Comment:
Can you catch a more specific exception than this? This can cause misleading errors on account of how broad Exception is in Python

Reply:
The things I really want to catch are:

if the import from tableauhyperapi.impl.dll fails, e.g. because we redesigned our internal package structure in HyperAPI
if any of the lib.hyper_decode_date properties does not exists (e.g. because we remove or rename a function)
if any of the asserts triggers
The easiest way to catch those 3 exception sources is to catch AssertionError, ImportError and AttributeError.

Is this sufficient, or would you prefer some other/more targeted error handling approach?

Reply:
Updated the error handling to be much more targeted